add workflow for compact 3 node cluster on vsphere

WenXinWei opened the pull request:
add workflow for compact 3 node cluster on vsphere 
epic:https://issues.redhat.com/browse/CORS-1597

WenXinWei commented:
make jobs

WenXinWei commented:
/retest

WenXinWei commented:
/test

WenXinWei commented:
/test all

WenXinWei commented:
/pj-rehearse

WenXinWei commented:
/pj-rehearse

WenXinWei commented:
@gpei  please help review.

gpei commented on ci-operator/step-registry/ipi/conf/vsphere/ipi-conf-vsphere-ref.yaml:
```diff
@@ -13,6 +13,10 @@ ref:
   - namespace: test-credentials
     name: ci-ibmcloud
     mount_path: /var/run/vault/ibmcloud
+  env:
```
How about keeping the same concept with other cloud providers like AWS to use `SIZE_VARIANT = compact` to indicate it's a compact 3-node cluster. 

https://github.com/openshift/release/blob/master/ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh#L143-L145

if [[ "${SIZE_VARIANT}" == "compact" ]], then we can set replicas of worker =0 and other master/worker parameters to `MACHINE_POOL_OVERRIDES` in a separate judgement statement.

WenXinWei replied:
good idea, I will work on this later.

WenXinWei replied:
@gpei test pass, please help review on the code.

WenXinWei commented:
/pj-rehearse

gpei commented on ci-operator/step-registry/ipi/conf/vsphere/ipi-conf-vsphere-commands.sh:
```diff
@@ -50,6 +50,27 @@ if [ ! -z ${VERSION} ]; then
+      osDisk:
+        diskSizeGB: 120"
+    fi
+    if [[ "${SIZE_VARIANT}" == "compact" ]]; then
```
@WenXinWei as we talked before, I think we don't need to put the `SIZE_VARIANT == compact` in the `if [ ! -z ${VERSION} ];` section, 
it should have no relationship with the VERSION parameter.

gpei commented on ci-operator/step-registry/ipi/conf/vsphere/ipi-conf-vsphere-commands.sh:
```diff
@@ -50,6 +50,27 @@ if [ ! -z ${VERSION} ]; then
+        diskSizeGB: 120"
+    fi
+    if [[ "${SIZE_VARIANT}" == "compact" ]]; then
+      MACHINE_POOL_OVERRIDES="controlPlane:
```
let's also print some useful log when running into `compact` case like:
```
echo "Compact SIZE_VARIANT was configured, setting worker's replicas to 0"
```

gpei commented on ci-operator/step-registry/cucushift/installer/rehearse/vsphere/ipi/compact-3node-cluster/cucushift-installer-rehearse-vsphere-ipi-compact-3node-cluster-workflow.yaml:
```diff
@@ -0,0 +1,9 @@
+    pre:
+    - chain: cucushift-installer-rehearse-vsphere-ipi-compact-3node-cluster-provision
+    post:
+    - chain: cucushift-installer-rehearse-vsphere-ipi-compact-3node-cluster-deprovision
```
I guess we can use `cucushift-installer-rehearse-vsphere-ipi-deprovision ` chain directly, so no need to add the `cucushift-installer-rehearse-vsphere-ipi-compact-3node-cluster-deprovision` chain.

WenXinWei replied:
ok， got it.

gpei commented:
BTW, let's update the name in workflow and chain from `vsphere-ipi-compact-3node-cluster` to `vsphere-ipi-compact` to be consistent with  https://github.com/openshift/release/pull/35212/files

WenXinWei commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.13-installer-rehearse-vsphere

WenXinWei commented:
@gpei  I have updated the code accordingly, job also passed with compact 3 node. please help review.

WenXinWei commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.13-installer-rehearse-vsphere

WenXinWei commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.13-installer-rehearse-vsphere

gpei commented:
/lgtm

WenXinWei commented:
@patrickdillon hi patrickdillon， please help review and approve the code changed in ipi-conf-vsphere-commands.sh

patrickdillon commented:
This LGTM but it would probably be best to get signoff from either @rvanderp3 or @jcpowermac

jcpowermac commented on ci-operator/step-registry/ipi/conf/vsphere/ipi-conf-vsphere-ref.yaml:
```diff
@@ -6,6 +6,20 @@ ref:
+      The sizes are:
+      * "" (default) - 4 vCPU, 16GB control plane nodes, default workers
+      * "compact" - 8 vCPU, 32GB control plane nodes, no workers
+      * "large" - 16 vCPU, 64GB+ control plane nodes, default workers, suitable for clusters up to 250 nodes
```
This concerns me a little. vSphere env(s) has finite resources using large or xlarge probably won't be an option any time soon.

rvanderp3 replied:
I second that concern.  The expectation is that 'default/compact' are the only valid options for vSphere.  I'll lgtm it, but bear in mind the resource constraints in our CI environments.

rvanderp3 replied:
oh, and it still needs a rebase :)

WenXinWei replied:
ok, the SIZE_VARIANT here we use in vsphere only set to "compact", because it's a common env, also used in other steps, "large" would not be used in vsphere. So maybe i can delete the note for "large"?

rvanderp3 replied:
yeah, let's do that.  thanks!

WenXinWei replied:
@jcpowermac @rvanderp3 please help review again.

jcpowermac commented:
Looks fine to me, needs a rebase though

WenXinWei commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.13-installer-rehearse-vsphere

WenXinWei commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.13-installer-rehearse-vsphere

WenXinWei commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.13-installer-rehearse-vsphere

WenXinWei commented:
@jianlinliu @jcpowermac  @rvanderp3 
Please help reivew the mr.  I moved SIZE_VARIANT from cucushift-installer-rehearse-vsphere-ipi-compact-provision-chain.yaml to cucushift-installer-rehearse-vsphere-ipi-compact-workflow.yaml, removed the cucushift-installer-rehearse-vsphere-ipi-compact-provision-chain.yaml and  just reused the cucushift-installer-rehearse-vsphere-ipi-provision. also remove the "large" note in ipi-conf-vsphere-ref.yaml.

WenXinWei commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.13-installer-rehearse-vsphere

WenXinWei commented on ci-operator/step-registry/ipi/conf/vsphere/ipi-conf-vsphere-commands.sh:
```diff
@@ -33,6 +33,22 @@ RESOURCE_POOL_DEF=""
 VERSION=$(echo "${JOB_NAME}" | grep -o -E '4\.[0-9]+')
 set -o errexit
+
+if [[ "${SIZE_VARIANT}" == "compact" ]]; then
```
ok, i have moved the block behind if [ ! -z ${VERSION} ]; section. now the version less than 4.11 also can affect by compact settings.

WenXinWei commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.13-installer-rehearse-vsphere

jianlinliu commented:
/pj-rehearse

jianlinliu commented:
/lgtm

WenXinWei commented:
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.13-installer-rehearse-vsphere

jianlinliu commented:
/lgtm

jianlinliu commented:
/pj-rehearse ack
